# Voltix task2 _ Python 

Name : Alaa Ayman Ramadan    
Raw Data : FactSale Dataset

Methods:
1. Import libraries , Check raw Data
2. Clean Data
3. Feature Engineering
4. EDA
5. Charts
6. KPIs
7. Report

## 0) Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


colors = ["#FFC349", "#525EA7", "#5FACD3", "#97DDE9", "#FF9D50"]
sns.set_palette(colors)

print("Libraries are successfuly imported")

Libraries are successfuly imported


## 1) Import and check raw data

In [2]:
try:
    df = pd.read_csv("FactSale.csv", encoding="utf-8")
    print('utf-8 is now running')
except UnicodeDecodeError:
    print("Problem in utf-8")
print('#'*50)
print("Shape of Data:", df.shape)
print('#'*50)

utf-8 is now running
##################################################
Shape of Data: (26397, 21)
##################################################


In [3]:
print('#'*50)
print('first and last 2 rows of raw data :')
print('#'*50)
pd.set_option('display.max_columns', None)
df.head(2)

##################################################
first and last 2 rows of raw data :
##################################################


,Sale Key,City Key,Customer Key,Bill To Customer Key,Stock Item Key,Invoice Date Key,Delivery Date Key,Salesperson Key,WWI Invoice ID,Description,Package,Quantity,Unit Price,Tax Rate,Total Excluding Tax,Tax Amount,Profit,Total Including Tax,Total Dry Items,Total Chiller Items,Lineage Key
0,49265,41568,0,0,204,10/22/2013,10/23/2013,83,15189,DBA joke mug - mind if I join you? (White),Each,6,13.0,15,78.0,11.7,51.0,89.7,6,0,11
1,49372,48937,0,0,173,10/22/2013,10/23/2013,83,15218,Developer joke mug - a foo walks into a bar (B...,Each,6,13.0,15,78.0,11.7,51.0,89.7,6,0,11


In [4]:
df.tail(2)

,Sale Key,City Key,Customer Key,Bill To Customer Key,Stock Item Key,Invoice Date Key,Delivery Date Key,Salesperson Key,WWI Invoice ID,Description,Package,Quantity,Unit Price,Tax Rate,Total Excluding Tax,Tax Amount,Profit,Total Including Tax,Total Dry Items,Total Chiller Items,Lineage Key
26395,224963,72760,23,1,220,5/14/2016,5/15/2016,170,69493,Novelty chilli chocolates 500g,Bag,12,14.5,10,174.0,17.4,69.0,191.4,0,12,11
26396,226629,66094,136,1,220,5/24/2016,5/25/2016,171,70005,Novelty chilli chocolates 500g,Bag,48,14.5,10,696.0,69.6,276.0,765.6,0,48,11


In [5]:
print('#'*50)
print('information about raw data :')
print('#'*50)
df.info()

##################################################
information about raw data :
##################################################
<class 'pandas.DataFrame'>
RangeIndex: 26397 entries, 0 to 26396
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Sale Key              26397 non-null  int64  
 1   City Key              26397 non-null  int64  
 2   Customer Key          26397 non-null  int64  
 3   Bill To Customer Key  26397 non-null  int64  
 4   Stock Item Key        26397 non-null  int64  
 5   Invoice Date Key      26397 non-null  str    
 6   Delivery Date Key     26384 non-null  str    
 7   Salesperson Key       26397 non-null  int64  
 8   WWI Invoice ID        26397 non-null  int64  
 9   Description           26397 non-null  str    
 10  Package               26397 non-null  str    
 11  Quantity              26397 non-null  int64  
 12  Unit Price            26397 non-null  float64
 13  T

In [6]:
print("#"*50)
print("Data Types:")
print("#"*50)
print(df.dtypes)
# problem in invoice and delivery date key (str) --> date

##################################################
Data Types:
##################################################
Sale Key                  int64
City Key                  int64
Customer Key              int64
Bill To Customer Key      int64
Stock Item Key            int64
Invoice Date Key            str
Delivery Date Key           str
Salesperson Key           int64
WWI Invoice ID            int64
Description                 str
Package                     str
Quantity                  int64
Unit Price              float64
Tax Rate                  int64
Total Excluding Tax     float64
Tax Amount              float64
Profit                  float64
Total Including Tax     float64
Total Dry Items           int64
Total Chiller Items       int64
Lineage Key               int64
dtype: object


In [7]:
print("#"*50)
print('missing report :')
print("#"*50)

missing_report = pd.DataFrame({
 'Missing Count':df.isnull().sum(),
 '%':df.isnull().sum()/len(df) * 100
    })

missing_report

##################################################
missing report :
##################################################


,Missing Count,%
Sale Key,0,0.000000
City Key,0,0.000000
Customer Key,0,0.000000
Bill To Customer Key,0,0.000000
Stock Item Key,0,0.000000
Invoice Date Key,0,0.000000
Delivery Date Key,13,0.049248
Salesperson Key,0,0.000000
WWI Invoice ID,0,0.000000
Description,0,0.000000


In [8]:
print("#"*50)
print("Duplicate Rows:")
print("#"*50)
duplicated_number= df.duplicated().sum()
print(f'We have mumber of {duplicated_number} duplicates')

##################################################
Duplicate Rows:
##################################################
We have mumber of 0 duplicates


In [9]:
def eda_function(df):
    print("#"*50)
    print("Object Statistics:")
    print("#"*50)
    print(df.describe(include=["object",'str']))
    print("#"*50)
    print("Numerical Statistics:")
    print("#"*50)
    print(df.describe())
    return df

In [10]:
eda_function(df.head(1))

##################################################
Object Statistics:
##################################################
       Invoice Date Key Delivery Date Key  \
count                 1                 1   
unique                1                 1   
top          10/22/2013        10/23/2013   
freq                  1                 1   

                                       Description Package  
count                                            1       1  
unique                                           1       1  
top     DBA joke mug - mind if I join you? (White)    Each  
freq                                             1       1  
##################################################
Numerical Statistics:
##################################################
       Sale Key  City Key  Customer Key  Bill To Customer Key  Stock Item Key  \
count       1.0       1.0           1.0                   1.0             1.0   
mean    49265.0   41568.0           0.0                   0.0 

,Sale Key,City Key,Customer Key,Bill To Customer Key,Stock Item Key,Invoice Date Key,Delivery Date Key,Salesperson Key,WWI Invoice ID,Description,Package,Quantity,Unit Price,Tax Rate,Total Excluding Tax,Tax Amount,Profit,Total Including Tax,Total Dry Items,Total Chiller Items,Lineage Key
0,49265,41568,0,0,204,10/22/2013,10/23/2013,83,15189,DBA joke mug - mind if I join you? (White),Each,6,13.0,15,78.0,11.7,51.0,89.7,6,0,11


In [11]:
def get_column_types(df):
    numeric_cols = df.select_dtypes(include=np.number).columns.tolist()
    datetime_cols = df.select_dtypes(include=['datetime', 'datetime64']).columns.tolist()
    categorical_cols = df.select_dtypes(include=['object', 'str']).columns.tolist()

    print("Numeric Columns:")
    print(numeric_cols)

    print("\nCategorical Columns:")
    print(categorical_cols)

    print("\ndate Columns:")
    print(datetime_cols)
    

In [12]:
get_column_types(df)

Numeric Columns:
['Sale Key', 'City Key', 'Customer Key', 'Bill To Customer Key', 'Stock Item Key', 'Salesperson Key', 'WWI Invoice ID', 'Quantity', 'Unit Price', 'Tax Rate', 'Total Excluding Tax', 'Tax Amount', 'Profit', 'Total Including Tax', 'Total Dry Items', 'Total Chiller Items', 'Lineage Key']

Categorical Columns:
['Invoice Date Key', 'Delivery Date Key', 'Description', 'Package']

date Columns:
[]


## 2) Clean Data
before cleaning we knew that :

1- no duplication 

2- invoice date , delivery date is string (we will make it datetime)

3- delivery date has 13 nulls (we will fill it **delivery=invoice + 1**)لاحظت ان التوصيل يبقي بعد يوم واحد فقط في جميع الصفوف

In [13]:
class Cleaner:
    def __init__(self, data):
        self.data = data.copy()

    def fix_columns_trim(self):
        self.data.columns = ( self.data.columns.str.strip() )
        print("trimmed successfully")
        return self

    def fix_dates(self, cols):
        for col in cols:
                self.data[col] = pd.to_datetime(self.data[col])
        print('we fixed dates successfully from string to date_time')
        return self

    def fill_delivery_date(self):
        self.data['Delivery Date Key'] = self.data['Delivery Date Key'].fillna( 
            self.data['Invoice Date Key'] + pd.Timedelta(days=1)
        )
        print("Delivery dates filled 13 rows successfully")
        return self

    def get_data(self):
        return self.data


cleaner = Cleaner(df)
df = (
    cleaner.fix_columns_trim()
    .fix_dates(["Invoice Date Key", "Delivery Date Key"])
    .fill_delivery_date()
    .get_data()
)

trimmed successfully
we fixed dates successfully from string to date_time
Delivery dates filled 13 rows successfully


In [14]:
def get_column_types(df):
    numeric_cols = df.select_dtypes(include=np.number).columns.tolist()
    datetime_cols = df.select_dtypes(include=['datetime', 'datetime64']).columns.tolist()
    categorical_cols = df.select_dtypes(include=['object', 'str']).columns.tolist()

    print("Numeric Columns:")
    print(numeric_cols)

    print("\nCategorical Columns:")
    print(categorical_cols)

    print("\ndate Columns:")
    print(datetime_cols)
    

In [15]:
get_column_types(df)

Numeric Columns:
['Sale Key', 'City Key', 'Customer Key', 'Bill To Customer Key', 'Stock Item Key', 'Salesperson Key', 'WWI Invoice ID', 'Quantity', 'Unit Price', 'Tax Rate', 'Total Excluding Tax', 'Tax Amount', 'Profit', 'Total Including Tax', 'Total Dry Items', 'Total Chiller Items', 'Lineage Key']

Categorical Columns:
['Description', 'Package']

date Columns:
['Invoice Date Key', 'Delivery Date Key']


In [16]:
print('#'*50)
print('information after cleaning data :')
print('#'*50)
df.info()

##################################################
information after cleaning data :
##################################################
<class 'pandas.DataFrame'>
RangeIndex: 26397 entries, 0 to 26396
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Sale Key              26397 non-null  int64         
 1   City Key              26397 non-null  int64         
 2   Customer Key          26397 non-null  int64         
 3   Bill To Customer Key  26397 non-null  int64         
 4   Stock Item Key        26397 non-null  int64         
 5   Invoice Date Key      26397 non-null  datetime64[us]
 6   Delivery Date Key     26397 non-null  datetime64[us]
 7   Salesperson Key       26397 non-null  int64         
 8   WWI Invoice ID        26397 non-null  int64         
 9   Description           26397 non-null  str           
 10  Package               26397 non-null  str           
 11  Quantit

In [17]:
print("#"*50)
print('missing report after cleaning:')
print("#"*50)

missing_report = pd.DataFrame({
 'Missing Count':df.isnull().sum(),
 '%':df.isnull().sum()/len(df) * 100
    })

missing_report

##################################################
missing report after cleaning:
##################################################


,Missing Count,%
Sale Key,0,0.0
City Key,0,0.0
Customer Key,0,0.0
Bill To Customer Key,0,0.0
Stock Item Key,0,0.0
Invoice Date Key,0,0.0
Delivery Date Key,0,0.0
Salesperson Key,0,0.0
WWI Invoice ID,0,0.0
Description,0,0.0


### Outliers using IQR

In [18]:
df.head(2)

,Sale Key,City Key,Customer Key,Bill To Customer Key,Stock Item Key,Invoice Date Key,Delivery Date Key,Salesperson Key,WWI Invoice ID,Description,Package,Quantity,Unit Price,Tax Rate,Total Excluding Tax,Tax Amount,Profit,Total Including Tax,Total Dry Items,Total Chiller Items,Lineage Key
0,49265,41568,0,0,204,2013-10-22,2013-10-23,83,15189,DBA joke mug - mind if I join you? (White),Each,6,13.0,15,78.0,11.7,51.0,89.7,6,0,11
1,49372,48937,0,0,173,2013-10-22,2013-10-23,83,15218,Developer joke mug - a foo walks into a bar (B...,Each,6,13.0,15,78.0,11.7,51.0,89.7,6,0,11


In [19]:
def flag_outliers(data, col):
    q1 = data[col].quantile(0.25)
    q3 = data[col].quantile(0.75)
    iqr = q3 - q1
    lower_fence = q1 - 1.5 * iqr
    upper_fence = q3 + 1.5 * iqr
    is_outlier = ~ data[col].between(lower_fence, upper_fence)
    return is_outlier


for col in ["Profit", "Tax Amount", "Total Including Tax", "Unit Price",'Quantity', 'Unit Price',]:
        df[col + "_outlier"] = flag_outliers(df, col)
        print(f"number of outliers in {col+'_outliers'}:", df[col + "_outlier"].sum())

number of outliers in Profit_outliers: 3101
number of outliers in Tax Amount_outliers: 2326
number of outliers in Total Including Tax_outliers: 2326
number of outliers in Unit Price_outliers: 2742
number of outliers in Quantity_outliers: 1589
number of outliers in Unit Price_outliers: 2742


## check negative values

In [20]:
numeric_cols = ['Quantity', 'Unit Price', 'Tax Rate', 'Total Excluding Tax','Tax Amount', 'Profit', 'Total Including Tax']
for c in numeric_cols:
    neg_count = (df[c] < 0).sum()
    print(f'{c}: min = {df[c].min()} ------> negative values = {neg_count}')

Quantity: min = 1 ------> negative values = 0
Unit Price: min = 0.66 ------> negative values = 0
Tax Rate: min = 10 ------> negative values = 0
Total Excluding Tax: min = 3.2 ------> negative values = 0
Tax Amount: min = 0.48 ------> negative values = 0
Profit: min = -645.0 ------> negative values = 566
Total Including Tax: min = 3.68 ------> negative values = 0


#profit values is 566 negative but this is nirmal (خساره)

## 3) Data Validation

In [21]:
import pandas as pd

def data_validation(df: pd.DataFrame) -> pd.DataFrame:
    print('#' * 50)
    print('Validation Report:')
    print('#' * 50)

    # 1) Tax Amount Check: Tax = Total Excluding Tax * (Tax Rate / 100)
    calc_tax = df['Total Excluding Tax'] * df['Tax Rate'] / 100.0
    tax_diff = (calc_tax - df['Tax Amount']).abs()
    print(f"1) Tax Amount check -> records with diff > 0.01: {(tax_diff > 0.01).sum()}")

    # 2) Total Including Tax Check: Total = Total Excluding Tax + Tax Amount
    calc_total = df['Total Excluding Tax'] + df['Tax Amount']
    total_diff = (calc_total - df['Total Including Tax']).abs()
    print(f"2) Total Including Tax check -> records with diff > 0.01: {(total_diff > 0.01).sum()}")

    # 3) Subtotal Check: Total Excluding Tax = Quantity * Unit Price
    calc_excl = df['Quantity'] * df['Unit Price']
    excl_diff = (calc_excl - df['Total Excluding Tax']).abs()
    print(f"3) Quantity x Unit Price check -> records with diff > 0.01: {(excl_diff > 0.01).sum()}")

    # 4) Item Breakdown Check: Total Dry Items + Total Chiller Items = Quantity
    items_diff = (df['Total Dry Items'] + df['Total Chiller Items'] - df['Quantity']).abs()
    print(f"4) Dry + Chiller = Quantity check -> records with diff: {(items_diff != 0).sum()}")

    # 5) Date Sequence Check: Delivery Date must be >= Invoice Date
    date_issue = (df['Delivery Date Key'] < df['Invoice Date Key']).sum()
    print(f"5) Delivery Date before Invoice Date -> records with issue: {date_issue}")

    # 6) Unregistered Customers Check: Count Customer Key == 0
    unknown_customers = (df['Customer Key'] == 0).sum()
    pct = (unknown_customers / len(df) * 100) if len(df) > 0 else 0
    print(f"6) Unregistered/Walk-in Customers (Key = 0): {unknown_customers} ({pct:.1f}%)")

    print('#' * 50)
    return df
    
df = data_validation(df)

##################################################
Validation Report:
##################################################
1) Tax Amount check -> records with diff > 0.01: 0
2) Total Including Tax check -> records with diff > 0.01: 0
3) Quantity x Unit Price check -> records with diff > 0.01: 0
4) Dry + Chiller = Quantity check -> records with diff: 0
5) Delivery Date before Invoice Date -> records with issue: 0
6) Unregistered/Walk-in Customers (Key = 0): 9077 (34.4%)
##################################################


#Customer Key = 0 ,Bill To Customer Key = 0 --> we will deal with it as a will be handled as a separate category 
#(Unregistered Customers) during analysis. 
#They were not deleted because this is not an error, but it must be documented.

## Business Questions

**Sales & Performance Over Time**

1-What is the monthly/yearly trend for Sales (Revenue) and Profit? Is there a clear seasonality pattern?
2-What are the top-performing months of the year in terms of sales volume and revenue?

**Product Analytics**

3. What are the top-selling products (Stock Items) by volume versus the most profitable ones? Are they the same products?
4. Which package type (Package) is most in demand?

**Customer Segmentation**

5. What is the sales distribution between registered customers (valid Customer Key) versus Walk-in customers, and what is the revenue impact of each group?


**Sales Team & Regional Performance**

6. Who is the top Salesperson in terms of total revenue generated and profit contributed?
7. Which cities (City Key) generate the highest sales revenue?

**Profitability Analysis**

8. What is the overall Profit Margin percentage, and which products or time periods operate at a net loss?

## cleaned data

In [22]:
df.to_csv("fact_sale_cleaned.csv", index=False)
print("Data cleaned done")

Data cleaned done
